_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,636 @@
+
+### Exhaustive case analysis
+
+> ...when you have eliminated all which is impossible,
+> then whatever remains, however improbable, must be the truth.
+>
+> -- Sherlock Holmes (Arthur Conan Doyle, "The Blanched Soldier")
+
+`match` enforces exhaustive case analysis, which helps catch bugs in
+program logic and ensures that the value of a match expression is
+well-defined.
+
+Case analysis is an important problem solving technique: Dividing a
+problem into some number of separate cases, and then analyzing each
+case individually.
+
+For this method of problem solving to work, the cases must be
+*collectively exhaustive*; otherwise, a case that was not covered
+would mean a potential problem instance for which no solution has been
+identified.
+
+This brings us to one of the fundamental restrictions of Rust's
+`match` construct: the collection of cases must be exhaustive. In
+other words, every possible input value must be covered by the pattern
+for a least one arm in the match.
+
+So, for example, the following code is rejected at compile-time.
+
+```rust
+fn suggest_guess_broken(prior_guess: u32, answer: Answer) {
+    let next_guess = match answer {
+        Answer::Higher => prior_guess + 10,
+        Answer::Lower  => prior_guess - 1,
+        // ERROR: non-exhaustive patterns: `Bingo` not covered
+    };
+    println!("maybe try {} next", next_guess);
+}
+```
+
+Many other languages offer a pattern matching construct (ML and
+various macro-based `match` implementations in Scheme both come to
+mind), but not all of them have this restriction.
+
+Rust has this restriction for two reasons:
+
+* First, as noted above, dividing a problem into cases only yields a
+general solution if the cases are exhaustive. Exhaustiveness-checking
+exposes logical errors.
+
+* Second, since `match` is an expression form, exhaustiveness ensures
+that such expressions always evaluates to a value of the correct type
+(or jump elsewhere in the program, as illustrated here):
+
+```rust
+fn suggest_guess_fixed(prior_guess: u32, answer: Answer) {
+    let next_guess = match answer {
+        Answer::Higher => prior_guess + 10,
+        Answer::Lower  => prior_guess - 1,
+        Answer::Bingo  => {
+            println!("we won with {}!", prior_guess);
+            return;
+        }
+    };
+    println!("maybe try {} next", next_guess);
+}
+
+#[test]
+fn demo_guess_fixed() {
+	suggest_guess_fixed(10, Answer::Higher);
+	suggest_guess_fixed(20, Answer::Lower);
+	suggest_guess_fixed(19, Answer::Bingo);
+}
+```
+
+The `suggest_guess_fixed` function illustrates that `match` can handle
+some cases early (and then immediately return from the function),
+while computing whatever values are needed from the remaining cases
+and letting them fall through to the remainder of the function
+body. We can add this special case handling via `match` without fear
+of overlooking a case, because `match` will enforce the case
+analysis to be exhaustive.
```

Comment:
Maybe it's worth talking about this as a refactoring aid? I certainly rely a lot on being able to add a variant to an enum and being assured I will be directed to all the places in my code that need to be updated.